File: apps/docker-server/cmd/renewlet/notification_dingtalk_test.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"net"
	"net/http"
	"net/url"
	"strings"
	"testing"
	"time"
)

func TestRenderWebhookPayloadTemplateEscapesMultilineContent(t *testing.T) {
	body, err := renderWebhookPayloadTemplate(`{"title":"{title}","content":"{content}","nested":["{timestamp}",{"copy":"{content}"}]}`, notificationMessage{
		Title:     "Renewlet",
		Content:   "即将续费：\n- GitHub：2026-08-01\n- Figma：2026-08-02",
		Timestamp: "2026-07-20 08:00 CST",
	}, localeZhCN)
	if err != nil {
		t.Fatal(err)
	}
	var payload map[string]interface{}
	if err := json.Unmarshal(body, &payload); err != nil {
		t.Fatalf("expected valid JSON body, got %q: %v", string(body), err)
	}
	if got := payload["content"]; got != "即将续费：\n- GitHub：2026-08-01\n- Figma：2026-08-02" {
		t.Fatalf("unexpected content %#v", got)
	}
	nested := payload["nested"].([]interface{})
	if nested[0] != "2026-07-20 08:00 CST" {
		t.Fatalf("unexpected nested timestamp %#v", nested)
	}
}

func TestRenderWebhookPayloadTemplateRejectsInvalidJSON(t *testing.T) {
	_, err := renderWebhookPayloadTemplate(`{"content": "{content}"`, notificationMessage{Content: "line\nline"}, localeZhCN)
	if err == nil {
		t.Fatal("expected invalid JSON template error")
	}
	if !strings.Contains(err.Error(), "JSON") {
		t.Fatalf("expected JSON parse error, got %q", err)
	}
}

func TestSignedDingTalkWebhookURLOverwritesTimestampAndSign(t *testing.T) {
	now := time.UnixMilli(1_774_225_234_567)
	got, err := signedDingTalkWebhookURL("https://oapi.dingtalk.com/robot/send?access_token=ding-token&timestamp=old&sign=old", "SECsecret", now)
	if err != nil {
		t.Fatal(err)
	}
	parsed, err := url.Parse(got)
	if err != nil {
		t.Fatal(err)
	}
	query := parsed.Query()
	if query.Get("timestamp") != "1774225234567" {
		t.Fatalf("unexpected timestamp %q", query.Get("timestamp"))
	}
	mac := hmac.New(sha256.New, []byte("SECsecret"))
	_, _ = mac.Write([]byte("1774225234567\nSECsecret"))
	wantSign := base64.StdEncoding.EncodeToString(mac.Sum(nil))
	if query.Get("sign") != wantSign {
		t.Fatalf("unexpected sign %q, want %q", query.Get("sign"), wantSign)
	}
	if strings.Contains(got, "old") {
		t.Fatalf("expected old signature query to be overwritten, got %q", got)
	}
}

func TestSendDingTalkPostsMarkdownPayloadAndRequiresErrCodeZero(t *testing.T) {
	withSafeOutboundResolver(t)
	var gotURL string
	var gotPayload dingTalkMarkdownRequest
	restore := withNotificationHTTPClient(t, serverChanRoundTripFunc(func(req *http.Request) (*http.Response, error) {
		gotURL = req.URL.String()
		if req.Method != http.MethodPost {
			t.Fatalf("expected POST, got %s", req.Method)
		}
		if err := json.NewDecoder(req.Body).Decode(&gotPayload); err != nil {
			t.Fatal(err)
		}
		return serverChanTestResponse(http.StatusOK, `{"errcode":0,"errmsg":"ok"}`), nil
	}))
	defer restore()

	settings := defaultAppSettings()
	settings.DingTalkWebhookURL = "https://oapi.dingtalk.com/robot/send?access_token=ding-token"
	settings.DingTalkKeyword = "Renewlet"
	err := sendDingTalk(settings, notificationMessage{
		Title:     "Renewlet 订阅提醒",
		Content:   "即将到期：\n- GitHub：2026-08-01",
		Timestamp: "2026-07-20 08:00 CST",
	})
	if err != nil {
		t.Fatal(err)
	}
	if !strings.HasPrefix(gotURL, "https://oapi.dingtalk.com/robot/send?access_token=ding-token") {
		t.Fatalf("unexpected DingTalk URL %q", gotURL)
	}
	if gotPayload.MsgType != "markdown" || gotPayload.Markdown.Title != "Renewlet 订阅提醒" {
		t.Fatalf("unexpected DingTalk payload %#v", gotPayload)
	}
	if !strings.Contains(gotPayload.Markdown.Text, "Renewlet") || !strings.Contains(gotPayload.Markdown.Text, "GitHub") {
		t.Fatalf("expected Renewlet marker and multiline content in payload, got %#v", gotPayload.Markdown.Text)
	}
}

func TestSendDingTalkAppliesTitleAndContentTemplates(t *testing.T) {
	withSafeOutboundResolver(t)
	var gotPayload dingTalkMarkdownRequest
	restore := withNotificationHTTPClient(t, serverChanRoundTripFunc(func(req *http.Request) (*http.Response, error) {
		if err := json.NewDecoder(req.Body).Decode(&gotPayload); err != nil {
			t.Fatal(err)
		}
		return serverChanTestResponse(http.StatusOK, `{"errcode":0,"errmsg":"ok"}`), nil
	}))
	defer restore()

	settings := defaultAppSettings()
	settings.DingTalkWebhookURL = "https://oapi.dingtalk.com/robot/send?access_token=ding-token"
	settings.DingTalkKeyword = "安全词"
	settings.DingTalkTitleTemplate = "{brand} · {title} · {unknown}"
	settings.DingTalkContentTemplate = "{keyword}\n{title}\n{content}\n{timestamp}\n{itemCount}\n{unknown}"
	err := sendDingTalk(settings, notificationMessage{
		Title:     "订阅提醒",
		Content:   "即将到期：\n- GitHub：{timestamp}",
		Timestamp: "2026-07-20 08:00 CST",
		Items:     []notificationContentItem{{}, {}},
	})
	if err != nil {
		t.Fatal(err)
	}
	if gotPayload.Markdown.Title != "Renewlet · 订阅提醒 · {unknown}" {
		t.Fatalf("unexpected DingTalk title %#v", gotPayload.Markdown.Title)
	}
	if !strings.Contains(gotPayload.Markdown.Text, "安全词\n订阅提醒") || !strings.Contains(gotPayload.Markdown.Text, "2") {
		t.Fatalf("expected rendered DingTalk template to include variables, got %#v", gotPayload.Markdown.Text)
	}
	if strings.Count(gotPayload.Markdown.Text, "安全词") != 1 || !strings.HasSuffix(gotPayload.Markdown.Text, "\n\nRenewlet") {
		t.Fatalf("expected exact keyword to stay single and missing brand marker to be appended, got %#v", gotPayload.Markdown.Text)
	}
	if !strings.Contains(gotPayload.Markdown.Text, "{unknown}") {
		t.Fatalf("expected unknown template variable to stay literal, got %#v", gotPayload.Markdown.Text)
	}
	if !strings.Contains(gotPayload.Markdown.Text, "GitHub：{timestamp}") {
		t.Fatalf("expected placeholders inside message content to stay literal, got %#v", gotPayload.Markdown.Text)
	}
}

func TestSendDingTalkEmptyTemplatesKeepDefaultPayloadOrder(t *testing.T) {
	withSafeOutboundResolver(t)
	var gotPayload dingTalkTextRequest
	restore := withNotificationHTTPClient(t, serverChanRoundTripFunc(func(req *http.Request) (*http.Response, error) {
		if err := json.NewDecoder(req.Body).Decode(&gotPayload); err != nil {
			t.Fatal(err)
		}
		return serverChanTestResponse(http.StatusOK, `{"errcode":0,"errmsg":"ok"}`), nil
	}))
	defer restore()

	settings := defaultAppSettings()
	settings.DingTalkWebhookURL = "https://oapi.dingtalk.com/robot/send?access_token=ding-token"
	settings.DingTalkMessageType = dingtalkMessageTypeText
	err := sendDingTalk(settings, notificationMessage{
		Title:     "订阅提醒",
		Content:   "即将到期：\n- GitHub：2026-08-01",
		Timestamp: "2026-07-20 08:00 CST",
	})
	if err != nil {
		t.Fatal(err)
	}
	want := "订阅提醒\n\n即将到期：\n- GitHub：2026-08-01\n\n2026-07-20 08:00 CST\n\nRenewlet"
	if gotPayload.Text.Content != want {
		t.Fatalf("expected default DingTalk content order with footer marker, got %#v", gotPayload.Text.Content)
	}
}

func TestEnsureDingTalkContentMarkersAppendsLowNoiseMarkers(t *testing.T) {
	tests := []struct {
		name    string
		content string
		keyword string
		want    string
	}{
		{
			name:    "lowercase keyword with existing brand stays at footer",
			content: "Renewlet 订阅提醒\n\n正文",
			keyword: "renewlet",
			want:    "Renewlet 订阅提醒\n\n正文\n\nrenewlet",
		},
		{
			name:    "exact brand keyword is not duplicated",
			content: "Renewlet 订阅提醒\n\n正文",
			keyword: "Renewlet",
			want:    "Renewlet 订阅提醒\n\n正文",
		},
		{
			name:    "existing custom keyword is not duplicated",
			content: "提醒\n\n安全词\n正文",
			keyword: "安全词",
			want:    "提醒\n\n安全词\n正文\n\nRenewlet",
		},
		{
			name:    "missing brand and keyword share one footer marker",
			content: "提醒\n\n正文",
			keyword: "自定义关键词",
			want:    "提醒\n\n正文\n\nRenewlet · 自定义关键词",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := ensureDingTalkContentMarkers(tt.content, tt.keyword); got != tt.want {
				t.Fatalf("unexpected DingTalk markers:\nwant: %q\n got: %q", tt.want, got)
			}
		})
	}
}

func TestSendDingTalkSignsTextPayloadAndRedactsSecretsOnBusinessFailure(t *testing.T) {
	withSafeOutboundResolver(t)
	var gotURL string
	var gotPayload dingTalkTextRequest
	restore := withNotificationHTTPClient(t, serverChanRoundTripFunc(func(req *http.Request) (*http.Response, error) {
		gotURL = req.URL.String()
		if err := json.NewDecoder(req.Body).Decode(&gotPayload); err != nil {
			t.Fatal(err)
		}
		return serverChanTestResponse(http.StatusOK, `{"errcode":310000,"errmsg":"keywords not in content SECsecret ding-token"}`), nil
	}))
	defer restore()

	settings := defaultAppSettings()
	settings.DingTalkWebhookURL = "https://oapi.dingtalk.com/robot/send?access_token=ding-token&timestamp=old&sign=old"
	settings.DingTalkSecret = "SECsecret"
	settings.DingTalkKeyword = "自定义关键词"
	settings.DingTalkMessageType = dingtalkMessageTypeText
	err := sendDingTalk(settings, notificationMessage{Title: "提醒", Content: "正文", Timestamp: "time"})
	if err == nil {
		t.Fatal("expected DingTalk business error")
	}
	if gotPayload.MsgType != "text" || !strings.HasSuffix(gotPayload.Text.Content, "\n\nRenewlet · 自定义关键词") {
		t.Fatalf("unexpected DingTalk text payload %#v", gotPayload)
	}
	if strings.Contains(gotURL, "old") || !strings.Contains(gotURL, "timestamp=") || !strings.Contains(gotURL, "sign=") {
		t.Fatalf("expected signed URL to overwrite old query, got %q", gotURL)
	}
	channelErr := notificationChannelErrorFrom(err)
	if channelErr == nil || channelErr.details == nil || channelErr.details.RawResponseText == nil {
		t.Fatalf("expected upstream details, got %#v", err)
	}
	for _, forbidden := range []string{"SECsecret", "ding-token"} {
		if strings.Contains(err.Error(), forbidden) || strings.Contains(*channelErr.details.RawResponseText, forbidden) {
			t.Fatalf("DingTalk error leaked %q: %q %#v", forbidden, err.Error(), *channelErr.details.RawResponseText)
		}
	}
	if !strings.Contains(err.Error(), "310000") || !strings.Contains(*channelErr.details.RawResponseText, "[redacted]") {
		t.Fatalf("expected errcode and redacted details, got %q %#v", err.Error(), *channelErr.details.RawResponseText)
	}
}

func TestRequireDingTalkSuccessFailsBusinessCodes(t *testing.T) {
	for _, body := range []string{
		`{"errcode":410100,"errmsg":"too many requests"}`,
		`{"errcode":40035,"errmsg":"缺少参数 json"}`,
		`{"errcode":400105,"errmsg":"不支持的 msgtype"}`,
	} {
		err := requireDingTalkSuccess(serverChanTestResponse(http.StatusOK, body), localeZhCN, "ding-token")
		if err == nil {
			t.Fatalf("expected DingTalk body %s to fail", body)
		}
		if !strings.Contains(err.Error(), "DingTalk") || !strings.Contains(err.Error(), "errcode=") {
			t.Fatalf("unexpected DingTalk error %q", err)
		}
	}
}

func withSafeOutboundResolver(t *testing.T) {
	t.Helper()
	previous := outboundURLResolver
	outboundURLResolver = func(_ string) ([]net.IPAddr, error) {
		return []net.IPAddr{{IP: net.ParseIP("93.184.216.34")}}, nil
	}
	t.Cleanup(func() {
		outboundURLResolver = previous
	})
}

File: apps/worker/src/notification-dingtalk.test.ts
import { createDefaultAppSettings } from "@renewlet/shared/settings-defaults";
import type { NotificationEmailMessage } from "@renewlet/shared/email-template";
import type { ApiAppSettings } from "@renewlet/shared/schemas/settings";
import { afterEach, describe, expect, it, vi } from "vitest";
import { NotificationChannelError } from "./notification-errors";
import { renderWebhookPayloadTemplate } from "./notification-channel-send";
import { sendDingTalk, signedDingTalkWebhookUrl } from "./notification-dingtalk";

vi.mock("./smtp", () => ({
  notificationSmtpConfig: vi.fn(),
  sendSmtpEmail: vi.fn(),
}));

const baseMessage: NotificationEmailMessage = {
  title: "Renewlet",
  content: "即将到期：\n- GitHub：2026-08-01",
  timestamp: "2026-07-20 08:00 CST",
  hasPayload: true,
  items: [],
};

function settings(overrides: Partial<ApiAppSettings>): ApiAppSettings {
  return {
    ...createDefaultAppSettings(),
    timezone: "UTC",
    notificationTimeLocal: "08:00" as ApiAppSettings["notificationTimeLocal"],
    ...overrides,
  };
}

function objectBody(init: RequestInit | undefined): Record<string, unknown> {
  if (typeof init?.body !== "string") throw new Error("expected JSON string body");
  const parsed = JSON.parse(init.body) as unknown;
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error("expected JSON object body");
  return parsed as Record<string, unknown>;
}

afterEach(() => {
  vi.restoreAllMocks();
  vi.unstubAllGlobals();
});

describe("Cloudflare DingTalk notification sender", () => {
  it("renders webhook JSON templates by replacing only string leaves", () => {
    const body = renderWebhookPayloadTemplate(
      `{"title":"{title}","content":"{content}","nested":["{timestamp}",{"copy":"{content}"}],"count":2}`,
      baseMessage,
      "zh-CN",
    );
    const parsed = JSON.parse(body) as { content: string; nested: [string, { copy: string }]; count: number };
    expect(parsed.content).toBe("即将到期：\n- GitHub：2026-08-01");
    expect(parsed.nested[0]).toBe("2026-07-20 08:00 CST");
    expect(parsed.nested[1].copy).toBe("即将到期：\n- GitHub：2026-08-01");
    expect(parsed.count).toBe(2);
  });

  it("rejects invalid webhook JSON templates", () => {
    expect(() => renderWebhookPayloadTemplate(`{"content":"{content}"`, baseMessage, "zh-CN")).toThrow("JSON");
  });

  it("overwrites DingTalk timestamp and sign query parameters", async () => {
    const endpoint = await signedDingTalkWebhookUrl(
      "https://93.184.216.34/robot/send?access_token=ding-token&timestamp=old&sign=old",
      "SECsecret",
      1_774_225_234_567,
    );

    expect(endpoint.searchParams.get("timestamp")).toBe("1774225234567");
    expect(endpoint.searchParams.get("sign")).toBeTruthy();
    expect(endpoint.toString()).not.toContain("old");
  });

  it("sends markdown payloads and treats errcode 0 as success", async () => {
    const fetchMock = vi.fn(async (_input: RequestInfo | URL, _init?: RequestInit) => new Response(`{"errcode":0,"errmsg":"ok"}`, { status: 200 }));
    vi.stubGlobal("fetch", fetchMock);

    await sendDingTalk(settings({
      dingtalkWebhookUrl: "https://93.184.216.34/robot/send?access_token=ding-token",
      dingtalkKeyword: "Renewlet",
    }), baseMessage, "zh-CN");

    expect(fetchMock).toHaveBeenCalledTimes(1);
    const [url, init] = fetchMock.mock.calls[0]!;
    expect(String(url)).toBe("https://93.184.216.34/robot/send?access_token=ding-token");
    expect(init?.method).toBe("POST");
    const payload = objectBody(init);
    expect(payload["msgtype"]).toBe("markdown");
    expect(payload["markdown"]).toMatchObject({
      title: "Renewlet",
      text: expect.stringContaining("GitHub"),
    });
    expect(payload["markdown"]).toMatchObject({
      text: "Renewlet\n\n即将到期：\n- GitHub：2026-08-01\n\n2026-07-20 08:00 CST",
    });
  });

  it("appends lowercase brand keyword as a footer marker instead of prepending it", async () => {
    const fetchMock = vi.fn(async (_input: RequestInfo | URL, _init?: RequestInit) => new Response(`{"errcode":0,"errmsg":"ok"}`, { status: 200 }));
    vi.stubGlobal("fetch", fetchMock);

    await sendDingTalk(settings({
      dingtalkWebhookUrl: "https://93.184.216.34/robot/send?access_token=ding-token",
      dingtalkKeyword: "renewlet",
    }), {
      ...baseMessage,
      title: "Renewlet 订阅提醒",
    }, "zh-CN");

    const payload = objectBody(fetchMock.mock.calls[0]?.[1]);
    expect(payload["markdown"]).toMatchObject({
      text: "Renewlet 订阅提醒\n\n即将到期：\n- GitHub：2026-08-01\n\n2026-07-20 08:00 CST\n\nrenewlet",
    });
  });

  it("applies DingTalk title and content templates while keeping unknown variables literal", async () => {
    const fetchMock = vi.fn(async (_input: RequestInfo | URL, _init?: RequestInit) => new Response(`{"errcode":0,"errmsg":"ok"}`, { status: 200 }));
    vi.stubGlobal("fetch", fetchMock);

    await sendDingTalk(settings({
      dingtalkWebhookUrl: "https://93.184.216.34/robot/send?access_token=ding-token",
      dingtalkKeyword: "安全词",
      dingtalkTitleTemplate: "{brand} · {title} · {unknown}",
      dingtalkContentTemplate: "{keyword}\n{title}\n{content}\n{timestamp}\n{itemCount}\n{unknown}",
    }), {
      ...baseMessage,
      title: "订阅提醒",
      content: "即将到期：\n- GitHub：{timestamp}",
      items: [
        { type: "renewal", name: "GitHub", price: "10", currency: "USD", status: "active", targetDate: "2026-08-01", reminderDays: 3 },
        { type: "trial", name: "Figma", price: "12", currency: "USD", status: "trial", targetDate: "2026-08-02", reminderDays: 3 },
      ],
    }, "zh-CN");

    const payload = objectBody(fetchMock.mock.calls[0]?.[1]);
    expect(payload["markdown"]).toMatchObject({
      title: "Renewlet · 订阅提醒 · {unknown}",
      text: expect.stringContaining("安全词\n订阅提醒"),
    });
    const text = (payload["markdown"] as { text: string }).text;
    expect(text.match(/安全词/g)).toHaveLength(1);
    expect(text.endsWith("\n\nRenewlet")).toBe(true);
    expect(JSON.stringify(payload["markdown"])).toContain("2");
    expect(JSON.stringify(payload["markdown"])).toContain("{unknown}");
    expect(JSON.stringify(payload["markdown"])).toContain("GitHub：{timestamp}");
  });

  it("signs text payloads and fails non-zero DingTalk business codes without leaking secrets", async () => {
    const fetchMock = vi.fn(async (_input: RequestInfo | URL, _init?: RequestInit) => new Response(
      `{"errcode":310000,"errmsg":"keywords not in content SECsecret ding-token"}`,
      { status: 200 },
    ));
    vi.stubGlobal("fetch", fetchMock);

    const caught = await sendDingTalk(settings({
      dingtalkWebhookUrl: "https://93.184.216.34/robot/send?access_token=ding-token&timestamp=old&sign=old",
      dingtalkSecret: "SECsecret",
      dingtalkKeyword: "自定义关键词",
      dingtalkMessageType: "text",
    }), { ...baseMessage, title: "提醒", content: "正文" }, "zh-CN").catch((error: unknown) => error);

    expect(fetchMock).toHaveBeenCalledTimes(1);
    const [url, init] = fetchMock.mock.calls[0]!;
    const sentUrl = new URL(String(url));
    expect(sentUrl.searchParams.get("timestamp")).toBeTruthy();
    expect(sentUrl.searchParams.get("sign")).toBeTruthy();
    expect(sentUrl.toString()).not.toContain("old");
    const payload = objectBody(init);
    expect(payload["msgtype"]).toBe("text");
    expect(payload["text"]).toMatchObject({
      content: "提醒\n\n正文\n\n2026-07-20 08:00 CST\n\nRenewlet · 自定义关键词",
    });

    expect(caught).toBeInstanceOf(NotificationChannelError);
    if (caught instanceof NotificationChannelError) {
      expect(caught.message).toContain("310000");
      expect(caught.message).not.toContain("SECsecret");
      expect(caught.message).not.toContain("ding-token");
      expect(caught.details?.rawResponseText).toContain("[redacted]");
      expect(caught.details?.rawResponseText).not.toContain("SECsecret");
      expect(caught.details?.rawResponseText).not.toContain("ding-token");
    }
  });

  it.each([410100, 40035, 400105])("fails DingTalk business errcode %s", async (errcode) => {
    const fetchMock = vi.fn(async () => new Response(JSON.stringify({ errcode, errmsg: "failed" }), { status: 200 }));
    vi.stubGlobal("fetch", fetchMock);

    await expect(sendDingTalk(settings({
      dingtalkWebhookUrl: "https://93.184.216.34/robot/send?access_token=ding-token",
    }), baseMessage, "zh-CN")).rejects.toThrow(`errcode=${errcode}`);
  });
});
